### DISCUSSION TITLE
Bump or height map for OglModel

### DISCUSSION THREAD

**Discussion by BrunoB81HK** [sofa-framework/sofa#2815] [Write / Run a simulation] [answered]
Hi,

My question is simple, is there any way we can apply a normal map on an `OglModel` so that the visual model will look textured?

I would like to have this :

![tumor (1)](https://user-images.githubusercontent.com/50412835/160522851-a1b250dd-c3e4-49b5-82f7-23570e619fbe.png)

instead of this :

![Capture d’écran 2022-03-28 224319](https://user-images.githubusercontent.com/50412835/160522381-d306aa6c-eac5-4e7d-a50b-04547d479c14.png)

Thanks a lot for your help!

Bruno

**Comment by fredroy** [comment]
Hello @BrunoB81HK !

It is possible to have the same effect in SOFA but it is using a kind-of experimental/not really documented things from SOFA, as we realised that we should delegate advanced rendering to the dedicated software (unity, vedo with py3, etc)
But yes, it is possible 🤠, here is an example of a textured liver with diffusion and normal maps among other things:
(this is only the visual node)
```xml


<Node name="Visual" >
   <OglShader vertFilename="shaders/shaderLibrary.glsl" fragFilename="shaders/shaderLibrary.glsl"  />
   <OglModel name="organs_liver_gall" filename="m_organs_Liver.obj" putOnlyTexCoords="true"  computeTangents="true" />
   <OglTexture id="DiffuseMap" textureFilename="organs_liver_gall.png" textureUnit="1" />
   <OglTexture id="NormalMap" textureFilename="organs_liver_gall_normalmap.png" textureUnit="2" />
   <OglShaderDefineMacro id="DiffuseMap_Present" />
   <OglShaderDefineMacro id="NormalMap_Present" />
   <OglFloat3Variable id="LightPosition" value="@../../../light/position.value" />
   <OglFloat3Variable id="LightColor" value="@../../../light/color.value" />
   <OglFloat3Variable id="DiffuseColor" value="1 1 1" />
   <OglFloat3Variable id="AmbientColor" value="0.3 0.3 0.3" />
   <OglFloat3Variable id="SpecularColor" value="1 1 1" />
   <OglFloatVariable  name="SpecularRoughness" value="0.05" />
   <OglFloatVariable  name="SpecularReflectance" value="0.28" />
    
   <BarycentricMapping name="visual mapping" input="@../mechanicalDofs" output="@organs_liver_gall" />
</Node>
```
This is using a shader code present in the `share/shaders` directory of SOFA.
Keep us in touch with some nice results (if you can 🥴)!

**Reply by BrunoB81HK** [reply]
I tried to implement this in my code but no texture is rendered. Also, the light is apparently not linking to my scene light. I hardcoded them to have a result.

My code currently consist of the following :

```xml
<Node name="Vis">
    <OglShader fileVertexShaders="['shaders/shaderLibrary.glsl']" fileFragmentShaders="['shaders/shaderLibrary.glsl']"  />
    <OglModel name="model" src="@../VisMesh" putOnlyTexCoords="true" computeTangents="true"/>
    <OglTexture id="DiffuseMap" textureFilename="../../../share/textures/LiverTumor.png" textureUnit="1"/>
    <OglTexture id="NormalMap" textureFilename="../../../share/textures/LiverTumorNormalMap.png" textureUnit="2"/>
    <OglShaderDefineMacro id="DiffuseMap_Present"/>
    <OglShaderDefineMacro id="NormalMap_Present"/>
    <OglFloat3Variable id="LightPosition" value="1 0 0" />
    <OglFloat3Variable id="LightColor" value="1 1 1" />
    <OglFloat3Variable id="DiffuseColor" value="1 1 1" />
    <OglFloat3Variable id="AmbientColor" value="0.3 0.3 0.3" />
    <OglFloat3Variable id="SpecularColor" value="1 1 1" />
    <OglFloatVariable  name="Roughness" value="0.15" />
    <OglFloatVariable  name="Specular" value="0.244" />

    <BarycentricMapping input="@../DOFs" output="@model"/>

</Node>
```

and I have some weird bugs like this :

![ezgif com-gif-maker](https://user-images.githubusercontent.com/50412835/161297447-bdd68eb1-b6e1-4e2f-aaad-2c580a48696b.gif)

**Reply by fredroy** [reply]
Yeah, does not look like bump mapping to me 🙃
Is there any warning from the shader compiler ?
Seems like some attributes (texcoord?) are not well linked...
Can you share the mesh and the texture files? I will **try** to make it work.

**Reply by BrunoB81HK** [reply]
Hi @fredroy, thanks a lot for your help! :slightly_smiling_face:

There was no warning when I ran it in SOFA. I must say that I use `runSofaGLFW` instead of `runSofa`. I tested with `runSofa` and I get the same results. 

Are the warnings from the shader compilers written when I compile everything with CMake or do I need to compile something else? If I don't, I don't think I saw any warning related to the shaders.

I will send you the needed files on Gitter.

**Comment by fredroy** [comment, marked as answer]
Hello @BrunoB81HK 
Actually, there was a problem in the OpenGL code https://github.com/sofa-framework/sofa/pull/2855
After fixing the types, the normal mapping is working.

![Capture](https://user-images.githubusercontent.com/11028016/162712480-89314f0b-cba6-4a35-b75c-dc0dec0ee0e0.PNG)


Seems a bit too much to me 🧐 some tweaks would be needed I guess

**Reply by BrunoB81HK** [reply]
TThis indeed looks too much. I will tweak it to my liking. :smile: 

Thanks a lot for your help and have a nice day!

**Reply by fredroy** [reply]
👍
